staging/vhost-device-gpu/src/device.rs

```diff
@@ -0,0 +1,1628 @@
+            .get_mut()
+            .get_queue_mut()
+            .iter(mem.clone())
+            .map_err(Error::CreateIteratorDescChain)?
+            .collect();
+
+        let mut signal_used_queue = false;
+        for desc_chain in desc_chains {
+            let head_index = desc_chain.head_index();
+            let mut reader = desc_chain
+                .clone()
+                .reader(&mem)
+                .map_err(Error::CreateReader)?;
+            let mut writer = desc_chain.writer(&mem).map_err(Error::CreateWriter)?;
+
+            self.process_queue_chain(
+                virtio_gpu,
+                vring,
+                head_index,
+                &mut reader,
+                &mut writer,
+                &mut signal_used_queue,
+            )?;
+        }
+
+        if signal_used_queue {
+            debug!("Notifying used queue");
+            vring
+                .signal_used_queue()
+                .map_err(Error::NotificationFailed)?;
+        }
+        debug!("Processing control queue finished");
+
+        Ok(())
+    }
+
+    fn handle_event(
+        &mut self,
+        device_event: u16,
+        virtio_gpu: &mut impl VirtioGpu,
+        vrings: &[VringRwLock],
+    ) -> IoResult<()> {
+        match device_event {
+            CONTROL_QUEUE | CURSOR_QUEUE => {
+                let vring = &vrings
+                    .get(device_event as usize)
+                    .ok_or_else(|| Error::HandleEventUnknown)?;
+                if self.event_idx {
+                    // vm-virtio's Queue implementation only checks avail_index
+                    // once, so to properly support EVENT_IDX we need to keep
+                    // calling process_queue() until it stops finding new
+                    // requests on the queue.
+                    loop {
+                        vring.disable_notification().unwrap();
+                        self.process_queue(virtio_gpu, vring)?;
+                        if !vring.enable_notification().unwrap() {
+                            break;
+                        }
+                    }
+                } else {
+                    // Without EVENT_IDX, a single call is enough.
+                    self.process_queue(virtio_gpu, vring)?;
+                }
+            }
+            POLL_EVENT => {
+                trace!("Handling POLL_EVENT");
+                virtio_gpu.event_poll()
+            }
+            _ => {
+                warn!("unhandled device_event: {}", device_event);
+                return Err(Error::HandleEventUnknown.into());
+            }
+        }
+
+        Ok(())
+    }
+
+    fn lazy_init_and_handle_event(
+        &mut self,
+        device_event: u16,
+        evset: EventSet,
+        vrings: &[VringRwLock],
+        _thread_id: usize,
+    ) -> IoResult<Option<EventFd>> {
+        // We use thread_local here because it is the easiest way to handle VirtioGpu being !Send
+        thread_local! {
```

Comment:
can't this be put in a LazyLock or so?

Reply:
I don't think so. Can you put `!Send` type in the  `LazyLock`?  
This doesn't seem to compile:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3df68bdc1f4e525d3e03710e147767f2

Another solution would be to just use `unsafe`  and have a normal static variable 😆, since this should never be executed by another thread. Or to spawn another thread just for the gpu object.

I am thinking of creating a proposal for a different API in the `vhost-user-backend` that will allow the user to spawn their own worker threads instead of framework spawning it for the user and a different mechanism for receiving the events, which would avoid situations like this and be more flexible.

Reply:
Oups, I somehow missed the !Send constraint. I guess wrapping the struct in another one that implement Send would do it as well.

Reply:
It would, but you have to do  **`unsafe`**`impl Send` for the wrapper struct. And keep in mind the `Rutabaga` struct inside this `RutabagaVirtioGpu` is `!Send` on purpose.  

https://github.com/google/crosvm/blob/18f43226d24e664b66a4ff5d6753e1c2619e64be/rutabaga_gfx/src/rutabaga_core.rs#L353-L355

More specifically this:
https://github.com/google/crosvm/blob/18f43226d24e664b66a4ff5d6753e1c2619e64be/rutabaga_gfx/src/virgl_renderer.rs#L336-L345

I believe using `unsafe` here would be correct here though, since we are in fact using `Rutabaga` from 1 thread only, it's just the Rust type system doesn't know that due to the way `vhost-user-backend` API is designed.

Reply:
yes, i know about having to write unsafe impl Send. The unsafe keyword is mostly here to get the user's attention about it, so they don't blindly implement for every struct.

Here is how I would it https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3df68bdc1f4e525d3e03710e147767f2

> I believe using unsafe here would be correct here though, since we are in fact using Rutabaga from 1 thread only, it's just the Rust type system doesn't know that due to the way vhost-user-backend API is designed.

What approach you had in mind when you say "using unsafe here"?

Reply:
@bilelmoussaoui  you're still doing same approach @mtjhrc explained in the first solution. probably you meant to do this instead? [https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=eb97d6be377f83e66a5b3cb4f6cb4b1d](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=eb97d6be377f83e66a5b3cb4f6cb4b1d)

Reply:
My bad, I meant to send this link https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c1bd07ed0e88030841140b00daa43b68 instead

Reply:
But at this point why use `LazyLock`, since we already assume (by our usage of unsafe) it is not being called from other threads, why lock it at this point?

Also btw not that when lazy initializing `RutabagaVirtioGpu` we also obtain the `event_poll_fd` and return it from this functions. (We only need to obtain the `event_poll_fd` once, so we do it when initializing the `RutabagaVirtioGpu`)

https://github.com/dorindabassey/vhost-device/blob/3e8940421207281fce46261cb9d22b6e87d37c1e/staging/vhost-device-gpu/src/device.rs#L566

Reply:
I think we can just use `static mut` something like this:
```Rust
fn lazy_init_and_handle_event(
    &mut self,
    device_event: u16,
    evset: EventSet,
    vrings: &[VringRwLock],
    _thread_id: usize,
) -> IoResult<Option<EventFd>> {
    debug!("Handle event called");
    if evset != EventSet::IN {
        return Err(Error::HandleEventNotEpollIn.into());
    };

    // We need to work around the issue that `RutabagaVirtioGpu` is !Send, and !Sync, but we
    // never access it from other threads, but the Rust type system doesn't know that.
    //
    // SAFETY: The VIRTIO_GPU static variable is only accessed from a specific worker thread,
    // It is not possible to alias the &mut reference to it, due to the way it is scoped.
    let gpu_ref: &mut Option<RutabagaVirtioGpu> = unsafe {
        static mut VIRTIO_GPU: Option<RutabagaVirtioGpu> = None;
        &mut *std::ptr::addr_of_mut!(VIRTIO_GPU)
    };

    let mut event_poll_fd = None;
    let virtio_gpu = match gpu_ref {
        Some(virtio_gpu) => virtio_gpu,
        None => {
            let gpu_backend = self.gpu_backend.take().ok_or_else(|| {
                io::Error::new(
                    ErrorKind::Other,
                    "set_gpu_socket() not called, GpuBackend missing",
                )
            })?;

            // We currently pass the CONTROL_QUEUE vring to RutabagaVirtioGpu, because we only
            // expect to process fences for that queue.
            let control_vring = &vrings[CONTROL_QUEUE as usize];

            // VirtioGpu::new can be called once per process (otherwise it panics),
            // so if somehow another thread accidentally wants to create another gpu here,
            // it will panic anyway
            let virtio_gpu = RutabagaVirtioGpu::new(control_vring, self.renderer, gpu_backend);
            event_poll_fd = virtio_gpu.get_event_poll_fd();

            gpu_ref.insert(virtio_gpu)
        }
    };

    self.handle_event(device_event, virtio_gpu, vrings)?;

    Ok(event_poll_fd)
}
```
(The safety comment can probably be improved)

Reply:
IMHO, I think the previous code handling using thread_local! is safer, atleast with `thread_local!` we don't need this unsafe block that `static mut` is introducing. I guess you are probably suggesting `static mut` because rutabaga is being called from a single thread.

Reply:
Well sure `thread_local!` should be the safest option. 

But what I am saying is that if we are willing to use `unsafe`, then we might as well use `static mut` instead of using `unsafe impl Send` + `LazyLock`.